src/types/complex-types/model.ts

```diff
@@ -284,37 +286,39 @@ function toPropertiesObject(declaredProps: ModelPropertiesDeclaration): ModelPro
       throw fail(
         "The default value of an attribute cannot be null or undefined as the type cannot be inferred. Did you mean `types.maybe(someType)`?"
       )
-      // its a primitive, convert to its type
-    } else if (isPrimitive(value)) {
-      return Object.assign({}, props, {
-        [key]: optional(getPrimitiveFactoryFromValue(value), value)
-      })
-      // map defaults to empty object automatically for models
-    } else if (value instanceof MapType) {
-      return Object.assign({}, props, {
-        [key]: optional(value, {})
-      })
+    }
+    // its a primitive, convert to its type
+    else if (isPrimitive(value)) {
+      props[key] = optional(getPrimitiveFactoryFromValue(value), value)
+    }
+    // map defaults to empty object automatically for models
+    else if (value instanceof MapType) {
+      props[key] = optional(value, {})
     } else if (value instanceof ArrayType) {
-      return Object.assign({}, props, { [key]: optional(value, []) })
-      // its already a type
-    } else if (isType(value)) {
-      return props
-      // its a function, maybe the user wanted a view?
-    } else if (devMode() && typeof value === "function") {
+      props[key] = optional(value, [])
+    }
+    // its already a type
+    else if (isType(value)) {
+      // do nothing, it's already a type
+    }
+    // its a function, maybe the user wanted a view?
+    else if (devMode() && typeof value === "function") {
       throw fail(
         `Invalid type definition for property '${key}', it looks like you passed a function. Did you forget to invoke it, or did you intend to declare a view / action?`
       )
-      // no other complex values
-    } else if (devMode() && typeof value === "object") {
+    }
+    // no other complex values
+    else if (devMode() && typeof value === "object") {
       throw fail(
         `Invalid type definition for property '${key}', it looks like you passed an object. Try passing another model type or a types.frozen.`
       )
-      // WTF did you pass in mate?
     } else {
       throw fail(
         `Invalid type definition for property '${key}', cannot infer a type from a value like '${value}' (${typeof value})`
       )
     }
+
+    return props
   }, declaredProps as any)
```

Comment:
I'm on mobile right now and can't check this, but we may want to spread this here (once) to avoid any potential accidental mutations at the call site.

If we can verify that this won't impact whatever is being passed into this function, then we wouldn't have to. But this would be my primary concern.

E.g., I'm worried that this could modify parent models when that isn't desired.

const Parent = t.model({ first: t.string })

const Child = Parent.props({ second: t.string })

... does Parent get modified here?

Reply:
@jamonholmgren - we are in the clear: https://github.com/mobxjs/mobx-state-tree/pull/2113/commits/fdda83954d5ef3b0806189b7b4f53ed083bd6595 

Also fixed a typo.